File: backend/app/schemas/document.py
from datetime import datetime
from uuid import UUID
from pydantic import BaseModel, ConfigDict, Field


class DocumentFolderCreate(BaseModel):
    name: str = Field(..., min_length=1)
    description: str | None = None
    parent_id: UUID | None = None
    allowed_roles: list[str] = Field(
        default=["ADMINISTRATOR", "DIRECTOR", "MANAGER", "RESIDENT"]
    )


class DocumentFolderUpdate(BaseModel):
    name: str | None = None
    description: str | None = None
    parent_id: UUID | None = None
    allowed_roles: list[str] | None = None


class DocumentFolderRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    name: str
    description: str | None = None
    parent_id: UUID | None = None
    allowed_roles: list[str] = []
    document_count: int = 0
    created_at: datetime
    updated_at: datetime


class DocumentFolderTreeRead(DocumentFolderRead):
    children: list["DocumentFolderTreeRead"] = []


class AssociationDocumentCreate(BaseModel):
    folder_id: UUID
    title: str = Field(..., min_length=1)
    description: str | None = None
    file_url: str = Field(..., min_length=1)
    file_size_bytes: int = Field(..., ge=0)
    mime_type: str = "application/pdf"
    publication_year: int | None = None
    publication_month: int | None = None
    tags: list[str] | None = None


class AssociationDocumentVersionCreate(BaseModel):
    title: str | None = None
    description: str | None = None
    file_url: str = Field(..., min_length=1)
    file_size_bytes: int = Field(..., ge=0)
    mime_type: str = "application/pdf"
    tags: list[str] | None = None


class AssociationDocumentRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    folder_id: UUID
    folder_name: str | None = None
    title: str
    description: str | None = None
    file_url: str
    file_size_bytes: int
    mime_type: str
    version_number: int
    previous_version_id: UUID | None = None
    publication_year: int | None = None
    publication_month: int | None = None
    tags: list[str] = []
    uploaded_by_id: UUID
    uploader_name: str | None = None
    created_at: datetime
    updated_at: datetime


class DocumentDownloadLogRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    document_id: UUID
    user_id: UUID
    user_name: str | None = None
    downloaded_at: datetime


class PaginatedDocumentRead(BaseModel):
    items: list[AssociationDocumentRead]
    total: int
    skip: int
    limit: int

File: frontend/src/types/document.ts
export interface DocumentFolder {
  id: string;
  name: string;
  description?: string | null;
  parent_id?: string | null;
  allowed_roles: string[];
  document_count: number;
  created_at: string;
  updated_at: string;
}

export interface DocumentFolderTree extends DocumentFolder {
  children: DocumentFolderTree[];
}

export interface AssociationDocument {
  id: string;
  folder_id: string;
  folder_name?: string | null;
  title: string;
  description?: string | null;
  file_url: string;
  file_size_bytes: number;
  mime_type: string;
  version_number: number;
  previous_version_id?: string | null;
  publication_year?: number | null;
  publication_month?: number | null;
  tags: string[];
  uploaded_by_id: string;
  uploader_name?: string | null;
  created_at: string;
  updated_at: string;
}

export interface DocumentDownloadLog {
  id: string;
  document_id: string;
  user_id: string;
  user_name?: string | null;
  downloaded_at: string;
}

export interface DocumentFolderCreatePayload {
  name: string;
  description?: string | null;
  parent_id?: string | null;
  allowed_roles?: string[];
}

export interface DocumentFolderUpdatePayload {
  name?: string | null;
  description?: string | null;
  parent_id?: string | null;
  allowed_roles?: string[] | null;
}

export interface AssociationDocumentCreatePayload {
  folder_id: string;
  title: string;
  description?: string | null;
  file_url: string;
  file_size_bytes: number;
  mime_type?: string;
  publication_year?: number | null;
  publication_month?: number | null;
  tags?: string[] | null;
}

export interface AssociationDocumentVersionCreatePayload {
  title?: string | null;
  description?: string | null;
  file_url: string;
  file_size_bytes: number;
  mime_type?: string;
  tags?: string[] | null;
}

export interface PaginatedDocumentResponse {
  items: AssociationDocument[];
  total: number;
  skip: number;
  limit: number;
}

export interface DocumentFilterParams {
  folder_id?: string;
  tag?: string;
  year?: number;
  month?: number;
  search?: string;
  skip?: number;
  limit?: number;
}
